code/modules/holodeck/computer.dm

```diff
@@ -110,207 +146,268 @@
 	. = TRUE
 	switch(action)
 		if("load_program")
-			var/program_to_load = text2path(params["type"])
-			if(!ispath(program_to_load))
-				return FALSE
-			var/valid = FALSE
+			var/program_to_load = params["id"]
+
 			var/list/checked = program_cache.Copy()
-			if(obj_flags & EMAGGED)
+			if (obj_flags & EMAGGED)
 				checked |= emag_programs
-			for(var/prog in checked)
-				var/list/P = prog
-				if(P["type"] == program_to_load)
+			var/valid = FALSE //dont tell security about this
+
+			for (var/prog in checked)//checks if program_to_load is any one of the loadable programs, if it isnt then it rejects it
+				var/list/check_list = prog
+				if (check_list["id"] == program_to_load)
 					valid = TRUE
 					break
-			if(!valid)
+			if (!valid)
 				return FALSE
-
-			var/area/A = locate(program_to_load) in GLOB.sortedAreas
-			if(A)
-				load_program(A)
+			//load the map_template that program_to_load represents
+			if(program_to_load)
+				load_program(program_to_load)
 		if("safety")
 			if((obj_flags & EMAGGED) && program)
 				emergency_shutdown()
-			nerf(obj_flags & EMAGGED)
+			nerf(obj_flags & EMAGGED,FALSE)
 			obj_flags ^= EMAGGED
-			say("Safeties restored. Restarting...")
+			say("Safeties reset. Restarting...")
+
+///this is what makes the holodeck not spawn anything on broken tiles (space and non engine plating / non holofloors)
+/datum/map_template/holodeck/update_blacklist(turf/placement, list/input_blacklist)
+	for (var/_turf in get_affected_turfs(placement))
+		var/turf/possible_blacklist = _turf
+		if (!istype(possible_blacklist, /turf/open/floor/holofloor))
+			if (istype(possible_blacklist, /turf/open/floor/engine))
+				continue
+			input_blacklist += possible_blacklist
+
+///loads the template whose id string it was given ("offline_program" loads datum/map_template/holodeck/offline)
+/obj/machinery/computer/holodeck/proc/load_program(map_id, force = FALSE, add_delay = TRUE)
+	if (program == map_id)
+		return
+
+	if (!is_operational)//load_program is called once with a timer (in toggle_power) we dont want this to load anything if its off
+		map_id = offline_program
+		force = TRUE
+
+	if (current_cd > world.time && !force)
+		say("ERROR. Recalibrating projection apparatus.")
+		return
+
+	if (add_delay)
+		current_cd = world.time + HOLODECK_CD
+		if(damaged)
+			current_cd += HOLODECK_DMG_CD
```

Comment:
Maybe add a new define for this.

Reply:
can you clarify what you mean by this?

Reply:
I mean something like this:
`#define COOLDOWN_ADD(cd_source, cd_index, cd_time) (cd_source.cd_index += cd_time)`

Reply:
completed already for you other reviews